backends/dpdk/dpdkArch.cpp

```diff
@@ -1665,22 +1662,56 @@ const IR::Node *CopyMatchKeysToSingleStruct::postorder(IR::KeyElement *element)
         insertions = it->second;
     }
 
-    auto keyName = element->expression->toString();
+    cstring keyName;
+    if (auto pe = element->expression->to<IR::PathExpression>()) {
+        keyName = pe->path->name;
+    } else if (auto mem = element->expression->to<IR::Member>()) {
+        /* We expect all nested structures are flattened and only
+         * expected expressions are 'h.<hdrname>.<fieldname> or m.<fieldname> */
+        if (auto mexpr = mem->expr->to<IR::Member>()) {
+            auto pe = mexpr->expr->to<IR::PathExpression>();
+            CHECK_NULL(pe);
+            BUG_CHECK(pe->path->name == "h",
+                      "Expected Member expressions of the form 'h.<hdrname>.<fieldname>");
+            keyName = "h." + mexpr->member + "." + mem->member;
+        } else if (auto mexpr = mem->expr->to<IR::PathExpression>()) {
+            BUG_CHECK(mexpr->path->name == "m",
+                      "Expected Member expressions of the form 'm.<fieldname>");
+            keyName = "m." + mem->member;
+        } else {
+            ::error(ErrorType::ERR_INVALID, "Invalid member expression '%1%' used as table key",
+                    mem);
+        }
+    } else if (auto m = element->expression->to<IR::MethodCallExpression>()) {
+        BUG_CHECK(isValidCall(m),
+                  "Method calls except isValid must be simplified before"
+                  " reaching here, found %1%",
+                  m->method);
+        /* Moving h.<hdr>.isValid() used as table key to Metadata */
+        auto mem = m->method->to<IR::Member>();
+        CHECK_NULL(mem);
+        keyName = "h." + mem->expr->to<IR::Member>()->member + "." + IR::Type_Header::isValid;
+    } else {
+        ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET, "Unsupported key expression %1%",
+                element->expression->toString());
+    }
+
+    if (keyName.isNullOrEmpty()) return element;
     bool isHeader = false;
     /* All header fields are prefixed with "h." and metadata fields are prefixed with "m."
      * Prefix the match field with control and table name */
     if (keyName.startsWith("h.")) {
         isHeader = true;
         keyName = keyName.replace('.', '_');
-        keyName =
-            keyName.replace("h_", control->name.toString() + "_" + table->name.toString() + "_");
+        keyName = keyName.replace(
```

Comment:
`originalName` can be null in specific scenarios. You can either check for that or use `toString()` here, which falls back to `name`. It depends on your expectations.

Reply:
The earlier code was using toString() which returns originalName if it is not null, name otherwise. In my attempt to remove toString(), I changed this as well :). In most cases, originalName for control and tables is not Null, I will add a check anyways. 
Thanks!

Reply:
Yes, just wanted to point out that the original usage was fine. For IR nodes `toString` can be unreliable but for `ID` `toString` should be stable. Although I have run into bugs because I expected `originalName` to be used but got `name`. Very subtle and unpleasant.

Reply:
Reverted to original implementation.